src/main/java/com/lucidworks/spark/SchemaPreservingSolrRDD.java

```diff
@@ -0,0 +1,537 @@
+  }
+
+  @Override
+  public JavaRDD<Row> toRows(final StructType schema, JavaRDD<SolrDocument> docs) {
+    JavaRDD<SolrDocument> rootdocs = docs.filter(new Function<SolrDocument, Boolean>() {
+        @Override
+        public Boolean call(SolrDocument solrDocument) throws Exception {
+            if (solrDocument.containsKey("__lwroot_s")) {
+                return true;
+            }
+            return false;
+        }
+    });
+    JavaPairRDD<String, SolrDocument> childdocs = docs.filter(new Function<SolrDocument, Boolean>() {
+      @Override
+      public Boolean call(SolrDocument solrDocument) throws Exception {
+        if (solrDocument.containsKey("__lwroot_s")) {
+          return false;
+        }
+        return true;
+      }
+    }).mapToPair(new PairFunction<SolrDocument, String, SolrDocument>() {
+      @Override
+      public Tuple2<String, SolrDocument> call(SolrDocument solrDocument){
+        return new Tuple2<String, SolrDocument>(solrDocument.get("id").toString(), solrDocument);
+      }
+    });
+    final Map<String, SolrDocument> childMap = childdocs.collectAsMap();
+    JavaRDD<Row> rows = rootdocs.map(new Function<SolrDocument, Row>() {
+      @Override
+      public Row call(SolrDocument solrDocument) throws Exception {
+        Row ret =  readData(solrDocument,  schema, collection, childMap);
+        return ret;
+      }
+    });
+    return rows;
+  }
+
+  public Row readData(SolrDocument doc, StructType st, String collection, Map<String, SolrDocument> childMap) throws IOException, SolrServerException {
+    ArrayList<Object> str = new ArrayList<Object>();
+    return recurseDataRead(doc , str, st, collection, childMap);
+  }
+
+  public Row recurseDataRead(SolrDocument doc, ArrayList<Object> x, StructType st, String collection, Map<String, SolrDocument> childMap) {
+    Boolean recurse = true;
+    java.util.Map<String, Object> x1 = doc.getFieldValueMap();
+    String[] x3 = st.fieldNames();
+    Object[] x2 = x1.keySet().toArray();
+    for (int i = 0; i < x2.length; i++) {
+      if (x2[i].toString().startsWith("links")) {
+        final String id = doc.get(x2[i].toString()).toString();
+        if (id != null) {
+          SolrDocument childDoc = childMap.get(id);
+          if (childDoc == null){
+            recurse = false;
+          }
+          if (recurse) {
+            //l = l + 1;
+            ArrayList<Object> str1 = new ArrayList<Object>();
+            x.add(recurseDataRead(childDoc, str1, (StructType) st.fields()[x.size()].dataType(), collection, childMap));
+          }
+        }
+      }
+      if (x3.length > x.size() && x1.get(x3[x.size()]+"_s") == null && !st.fields()[x.size()].dataType().typeName().equals("struct")){
+        x.add(null);
+      }
+      if ((x2[i].toString().substring(x2[i].toString().length() - 2, x2[i].toString().length()).equals("_s") && !x2[i].toString().startsWith("__lw") && !x2[i].toString().startsWith("links"))) {
```

Comment:
whoa! that's a lot of toString()'s in there! and many seem repeated ... consider just doing it once and using throughout ... 

Also, what's the story with timestamp and date?

Lastly, if I was writing this, I'd probably use reflection to map type to a convertToX method and then cache that ... but that's mostly stylistic